src/app/fdctl/run/tiles/fd_repair.c

```diff
@@ -0,0 +1,656 @@
+  peer_addr.l = 0;
+  peer_addr.addr = ip;
+  peer_addr.port = fd_ushort_bswap( hdr->udp->net_sport );
+
+  fd_repair_recv_packet( ctx->repair, ctx->buffer + hdr_sz, ctx->buffer_sz - hdr_sz, &peer_addr );
+}
+
+static inline void
+after_credit( void *             _ctx,
+              fd_mux_context_t * mux FD_PARAM_UNUSED ) {
+  fd_repair_tile_ctx_t * ctx = (fd_repair_tile_ctx_t *)_ctx;
+
+  fd_mcache_seq_update( ctx->net_out_sync, ctx->net_out_seq );
+
+  fd_repair_settime( ctx->repair, fd_log_wallclock() );
+  fd_repair_continue( ctx->repair );
+}
+
+static void
+privileged_init( fd_topo_t *      topo,
+                 fd_topo_tile_t * tile,
+                 void *           scratch ) {
+  (void)topo;
+  (void)tile;
+
+  FD_SCRATCH_ALLOC_INIT( l, scratch );
+  fd_repair_tile_ctx_t * ctx = FD_SCRATCH_ALLOC_APPEND( l, alignof(fd_repair_tile_ctx_t), sizeof(fd_repair_tile_ctx_t) );
+
+  uchar const * identity_key = fd_keyload_load( tile->sign.identity_key_path, /* pubkey only: */ 0 );
+  fd_memcpy( ctx->identity_private_key, identity_key, sizeof(fd_pubkey_t) );
+  fd_memcpy( ctx->identity_public_key.uc, identity_key + 32UL, sizeof(fd_pubkey_t) );
+
+  ctx->repair_config.private_key = ctx->identity_private_key;
+  ctx->repair_config.public_key  = &ctx->identity_public_key;
+}
+
+static void
+unprivileged_init( fd_topo_t *      topo,
+                   fd_topo_tile_t * tile,
+                   void *           scratch ) {
+
+  if( FD_UNLIKELY( tile->in_cnt != 4 ||
+                   strcmp( topo->links[ tile->in_link_id[ NET_IN_IDX     ] ].name, "net_repair")     ||
+                   strcmp( topo->links[ tile->in_link_id[ CONTACT_IN_IDX ] ].name, "gossip_repair" ) ||
+                   strcmp( topo->links[ tile->in_link_id[ STAKE_IN_IDX ] ].name,   "stake_out" )     ||
+                   strcmp( topo->links[ tile->in_link_id[ STORE_IN_IDX ] ].name,   "store_repair" ) ) ) {
+    FD_LOG_ERR(( "Repair tile has none or unexpected input links %lu %s %s",
+                 tile->in_cnt, topo->links[ tile->in_link_id[ 0 ] ].name, topo->links[ tile->in_link_id[ 1 ] ].name ));
+  }
+
+  if( FD_UNLIKELY( tile->out_cnt != 1 ||
+                   strcmp( topo->links[ tile->out_link_id[ STORE_OUT_IDX ] ].name, "repair_store" ) ) ) {
+    FD_LOG_ERR(( "repair tile has none or unexpected output links %lu %s %s",
+                 tile->out_cnt, topo->links[ tile->out_link_id[ 0 ] ].name, topo->links[ tile->out_link_id[ 1 ] ].name ));
+  }
+
+  if( FD_UNLIKELY( tile->out_link_id_primary == ULONG_MAX ) )
+    FD_LOG_ERR(( "repair tile has no primary output link" ));
+
+  /* Scratch mem setup */
+
+  FD_SCRATCH_ALLOC_INIT( l, scratch );
+  fd_repair_tile_ctx_t * ctx = FD_SCRATCH_ALLOC_APPEND( l, alignof(fd_repair_tile_ctx_t), sizeof(fd_repair_tile_ctx_t) );
+  ctx->repair = FD_SCRATCH_ALLOC_APPEND( l, fd_repair_align(), fd_repair_footprint() );
+
+  void * smem = FD_SCRATCH_ALLOC_APPEND( l, fd_scratch_smem_align(), fd_scratch_smem_footprint( SMAX ) );
```

Comment:
Can you just inline these constants here and document them. Who is using the scratch, and why do you know it's going to be large enough to serve the consumers?

Reply:
(SMAX, SDEPTH)

Reply:
Are you suggesting that i have local constants here because these are used in the footprint as well. Will add a note explaining how the scratch space is used